WalletWasabi/Crypto/Mac.cs

```diff
@@ -0,0 +1,56 @@
+using System;
+using System.Text;
+using NBitcoin.Secp256k1;
+using WalletWasabi.Crypto.Groups;
+using WalletWasabi.Helpers;
+
+namespace WalletWasabi.Crypto
+{
+	public class MAC : IEquatable<MAC>
+	{
+		private MAC(Scalar t, GroupElement U, GroupElement V)
+		{
+			this.t = CryptoGuard.NotZero(nameof(t), t);
+			this.U = CryptoGuard.NotNullOrInfinity(nameof(U), U);
+			this.V = CryptoGuard.NotNullOrInfinity(nameof(V), V);
+		}
+
+		public Scalar t { get; }
+		public GroupElement U { get; }
+		public GroupElement V { get; }
+
+		public static bool operator == (MAC a, MAC b) => a.Equals(b);
+		public static bool operator != (MAC a, MAC b) => !a.Equals(b);
+
+		public bool Equals(MAC? other) =>
+			(this, other) switch
+			{
+				(null, null) => true,
+				(null, {}) => false,
+				({}, null) => false,
+				({} a , {} b) when Object.ReferenceEquals(a, b) => true, 
+				({} a , {} b) => (a.t, a.U, a.V) == (b.t, b.U, b.V)
+			};
+
+		public override bool Equals(object? obj) =>
+			Equals(obj as MAC);
+
+		public override int GetHashCode() =>
+			HashCode.Combine(t, U, V).GetHashCode();
+
+		public static MAC ComputeMAC(CoordinatorSecretKey sk, GroupElement Ma, Scalar t)
+		{
+			Guard.NotNull(nameof(sk), sk);
+			CryptoGuard.NotNullOrInfinity(nameof(Ma), Ma);
+
+			var numsGeneratorSeed = Encoding.UTF8.GetString(t.ToBytes());
```

Comment:
i don't think this is right, since not all bytestrings are valid utf8.

i would prefer if `FromText` had a `byte[]` overload that was the "real" version, or even if it didn't support a string variant at all since SHA-256 is defined over bytestrings not unicode strings.

this also raises the question of whether or not the constructor should even support setting `U`, or if that should always be implicit based on `t`. i think i prefer doing that since i can see no reason to support setting `U` and `t` independently.